lib/edgehog_device/device.c

```diff
@@ -41,181 +41,333 @@ static edgehog_result_t add_interfaces(astarte_device_handle_t astarte_device);
+        return;
+    }
+    astarte_result_t ares = astarte_device_destroy(edgehog_device->astarte_device);
+    if (ares != ASTARTE_RESULT_OK) {
+        edgehog_device->astarte_error = ares;
+        EDGEHOG_LOG_ERR("Astarte device destroy error: %s", astarte_result_to_name(ares));
+    }
     edgehog_telemetry_destroy(edgehog_device->telemetry);
     free(edgehog_device);
 }
 
 edgehog_result_t edgehog_device_start(edgehog_device_handle_t edgehog_device)
 {
-    edgehog_initial_publish(edgehog_device);
-
-    edgehog_result_t res = edgehog_telemetry_start(edgehog_device);
-    if (res != EDGEHOG_RESULT_OK) {
-        EDGEHOG_LOG_ERR("Unable to start Edgehog device");
-    }
-
-    return res;
-}
-
-edgehog_result_t edgehog_device_stop(edgehog_device_handle_t edgehog_device, k_timeout_t timeout)
-{
-    edgehog_result_t res = edgehog_telemetry_stop(edgehog_device->telemetry, timeout);
-    if (res != EDGEHOG_RESULT_OK) {
-        EDGEHOG_LOG_ERR("Unable to stop the Edgehog device within the timeout");
+    astarte_result_t ares = astarte_device_connect(edgehog_device->astarte_device);
+    if (ares != ASTARTE_RESULT_OK) {
+        edgehog_device->astarte_error = ares;
+        EDGEHOG_LOG_ERR("Astarte device connection error: %s", astarte_result_to_name(ares));
+        return EDGEHOG_RESULT_ASTARTE_ERROR;
     }
-    return res;
+    edgehog_device->state = EDGEHOG_DEVICE_STARTED;
+    return EDGEHOG_RESULT_OK;
 }
 
-void edgehog_device_datastream_object_events_handler(
-    edgehog_device_handle_t edgehog_device, astarte_device_datastream_object_event_t event)
+edgehog_result_t edgehog_device_poll(edgehog_device_handle_t edgehog_device)
 {
-    if (!edgehog_device) {
-        EDGEHOG_LOG_ERR("Unable to handle event, Edgehog device undefined");
-        return;
+    astarte_result_t ares = astarte_device_poll(edgehog_device->astarte_device);
+    if (ares != ASTARTE_RESULT_OK) {
+        edgehog_device->astarte_error = ares;
+        EDGEHOG_LOG_ERR("Astarte device poll failure.");
+        return EDGEHOG_RESULT_ASTARTE_ERROR;
     }
 
-    astarte_device_data_event_t rx_event = event.data_event;
-
-    if ((strcmp(rx_event.interface_name, io_edgehog_devicemanager_OTARequest.name) == 0)
-        && (strcmp(rx_event.path, "/request") == 0)) {
-        edgehog_result_t ota_result = edgehog_ota_event(edgehog_device, &event);
-        if (ota_result != EDGEHOG_RESULT_OK) {
-            EDGEHOG_LOG_ERR("Unable to handle OTA update request");
+    edgehog_result_t eres = EDGEHOG_RESULT_OK;
+    if (edgehog_device->state == EDGEHOG_DEVICE_CONNECTED) {
+        if (!edgehog_device->initial_publish) {
+            edgehog_initial_publish(edgehog_device);
+            edgehog_device->initial_publish = true;
+        }
+        if (!edgehog_device->telemerty_service) {
+            eres = edgehog_telemetry_start(edgehog_device);
+            if (eres == EDGEHOG_RESULT_OK) {
+                edgehog_device->telemerty_service = true;
```

Comment:
Why do you define a boolean for retrieving this info? In the telemetry module we have an atomic bit you can test, so you can define a method like this: 
``` c
bool edgehog_telemetry_is_running(edgehog_device_handle_t device){
    edgehog_telemetry_t *telemetry = device->telemetry;
    if (!telemetry) {
        return false;
    }

    return atomic_test_bit(&telemetry->msgq_thread_state, MSGQ_THREAD_RUNNING_BIT);
}
```
What do you think? Also you don't have to update the `telemerty_service` field each time state is changed.

Reply:
Yes, that makes more sense